Python/Product/Analysis/LanguageServer/ParseQueue.cs

```diff
@@ -32,181 +32,225 @@ class ParseQueue {
+                _queue._parsingInProgress.Decrement();
+            }
+
+            public ParseTask ContinueAfter(ParseTask currentTask) {
+                // Set our previous task
+                Volatile.Write(ref _previous, currentTask?.Task);
+
+                return this;
+            }
+
+            public Task<IAnalysisCookie> Start() {
+                int actualState = Interlocked.CompareExchange(ref _state, QUEUED, UNSTARTED);
+                if (actualState != UNSTARTED) {
+                    if (actualState == DISPOSED) {
+                        throw new ObjectDisposedException(GetType().FullName);
                     }
-                } else if (doc is IPythonProjectEntry pyEntry) {
-                    bool complete = false;
-                    pyEntry.GetTreeAndCookie(out _, out var lastCookie);
-                    var lastVc = lastCookie as VersionCookie;
+                    throw new InvalidOperationException("cannot start parsing");
+                }
+
+                // If we were pending, calling start ensures we are queued
+                var previous = Interlocked.Exchange(ref _previous, null);
+                if (previous != null) {
                     try {
-                        var buffers = new SortedDictionary<int, BufferVersion>();
-                        foreach (var part in doc.DocumentParts.Reverse()) {
-                            using (var r = doc.ReadDocumentBytes(part, out var version)) {
-                                if (r == null) {
-                                    continue;
-                                }
-                                if (version >= 0 && lastVc != null && lastVc.Versions.TryGetValue(part, out var lastParse) && lastParse.Version >= version) {
-                                    buffers[part] = lastParse;
-                                    continue;
-                                }
-                                ParsePython(r, pyEntry, languageVersion, out var tree, out List<Diagnostic> diags);
-                                buffers[part] = new BufferVersion(
-                                    version,
-                                    tree,
-                                    diags.MaybeEnumerate()
-                                );
-                            }
-                        }
-                        if (!buffers.Any()) {
-                            throw new FileNotFoundException($"failed to parse file {pyEntry.DocumentUri.AbsoluteUri}", pyEntry.FilePath);
-                        }
-                        complete = true;
-                        result = UpdateTree(pyEntry, buffers);
-                    } finally {
-                        if (!complete) {
-                            AbortParsingTree(pyEntry);
-                        }
+                        previous.ContinueWith(StartAfterTask);
+                        return Task;
+                    } catch (ObjectDisposedException) {
```

Comment:
`ContinueWith` doesn't throw `ObjectDisposedException` in .Net 4.5 or later, even if you have disposed a `Task` (which you actually shouldn't: https://blogs.msdn.microsoft.com/pfxteam/2012/03/25/do-i-need-to-dispose-of-tasks/).

Reply:
They need to update the documentation :)